### PR TITLE
OIDC User Login flow

### LINKED ISSUE #54201: Feature request: Support Oauth2 authorization flow in kubectl
We would like  to see kubectl support OAuth token authorization flow. Essentially user can set necessary Endpoint, ClientID, ClientSecret and Scopes for Oauth. Using that information kubectl can initiate a Oauth authorization flow using user's local browser. Then use the issued token as token for Kubernetes authorization. Obviously this will not work in unattended modes. This can be very handy for OAuth providers that issue short lived tokens.

I have done a similar implementation using Google apis here: https://github.com/appscode/guard/blob/master/cmds/token.go

/kind feature
/sig cli

### PR DESCRIPTION
**Pull request by easeway**
**What this PR does / why we need it**:

Add a new login command to kubectl: `kubectl alpha login` which orchestrates interactive user login flow with a local web browser. It provides a convenient way without figuring out id-token and refresh-token with some other mechanisms and copies the values into kubeconfig manually.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #54201

**Special notes for your reviewer**:

This PR is submitted initially as a proposal. Please provide feedback, and I will help to follow up.

**Release note**:

```release-note
The following additional config properties are added in kubeconfig under auth-provider:

...
user:
- name: oidc-user
  user:
    auth-provider:
      client-id: client-id-from-provider
      client-secret: client-secret-from-provider
      idp-issuer-url: provider-IDP-URL
      callback-port: '8172'
      callback-path: /authcode
      extra-scopes: profile,email
      auth-options: prompt=consent

The additional properties are all optional:

- callback-port: specifies which port to listen on the local web server launched by kubectl for OIDC callback, must be consistent with the service registered in provider
- callback-path: the path in callback url, must be consistent with service registered in provider
- extra-scopes: scopes in addition to openid
- auth-options: parameters in the initial auth request, semi-comma separated

With this command, initially we don't need to specify id-token and refresh-token in the config.
Run this command, and once the login completes, id-token/refresh-token will be filled in.

One known issue with https://accounts.google.com, as google only issue refresh-token the first time user logins with the service, auth-options: prompt=consent is needed if refresh-token is expected to be updated in config every time.
```

### CODE REVIEW

**Comment by tamalsaha** [comment]
Thank you for doing this @easeway  ! I would like to say that OIDC should not be required for this. Since `kubectl` can work with `--token`, this should work for any OAuth2 provider (not just the ones who also send id token). How the auth token gets associated with an identity should be up to the `webhook authN server`.

**Comment by tamalsaha** [comment]
@kubernetes/sig-cli-pr-reviews

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/login.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+		if r.URL.Query().Get("state") != state {
+			http.Error(w, "invalid session", http.StatusBadRequest)
+			return
+		}
+		token, err := oauth2Conf.Exchange(ctx, r.URL.Query().Get("code"))
+		if err != nil {
+			http.Error(w, "failed to exchange token: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
+		idToken, ok := token.Extra("id_token").(string)
```
Comment (tamalsaha) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/login.go]: I would like to say that OIDC should not be required for this. Since `kubectl` can work with `--token`, this should work for any OAuth2 provider (not just the ones who also send id token). How the auth token gets associated with an identity should be up to the `webhook authN server`.

**Comment by liggitt** [comment]
@kubernetes/sig-cli-proposals @kubernetes/sig-auth-proposals

**Comment by liggitt** [comment]
cc @mikedanese

File: Godeps/Godeps.json [hunk 2]
```diff
@@ -2997,6 +3013,21 @@
 			"Rev": "20b71e5b60d756d3d2f80def009790325acc2b23"
 		},
 		{
+			"ImportPath": "gopkg.in/square/go-jose.v2",
```
Comment (mikedanese) [inline Godeps/Godeps.json]: go-jose is already vendored at github.com/square/go-jose. Can you just update that?

File: staging/src/k8s.io/client-go/rest/plugin.go [hunk 3]
```diff
@@ -30,8 +30,8 @@ type AuthProvider interface {
 	// WrapTransport allows the plugin to create a modified RoundTripper that
 	// attaches authorization headers (or other info) to requests.
 	WrapTransport(http.RoundTripper) http.RoundTripper
-	// Login allows the plugin to initialize its configuration. It must not
-	// require direct user interaction.
+	// Login allows the plugin to initialize its configuration. It may involve
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/rest/plugin.go]: This is a major change that hasn't been discussed.

**Comment by pwittrock** [comment]
cc @monopole

**Comment by pwittrock** [comment]
@easeway Thanks for your contribution.  Have you read the [SIG cli contributing guide](https://github.com/kubernetes/community/blob/master/sig-cli/CONTRIBUTING.md)?

File: staging/src/k8s.io/client-go/rest/plugin.go (hunk 3 above)
Comment (ericchiang) [reply staging/src/k8s.io/client-go/rest/plugin.go]: Other auth providers already require user interaction. e.g. the azure auth-n plugin can prompt you.

**Comment by ericchiang** [comment]
Beyond the sig-cli concerns I have a couple comments:

* @tamalsaha's right that this is really close to a pure OAuth2 login (we just grab the "id_token" instead of the "access_token" in the response). We might consider adding a mode for a regular access_token.
* @liggitt's brought up concerns in the past about transmitting tokens over HTTP, even via localhost. This is something we should at the very least call out.

Going to add this to tomorrow's SIG auth meeting so we can discuss this.

File: staging/src/k8s.io/client-go/rest/plugin.go (hunk 3 above)
Comment (mikedanese) [reply staging/src/k8s.io/client-go/rest/plugin.go]: So is the comment on the interface wrong or the azure implementation wrong?

Comment (ericchiang) [reply staging/src/k8s.io/client-go/rest/plugin.go]: The Login command has never been implemented, the Azure transport wrapper does this, so I'd call the comment is outdated.

Conversation about interactive prompts here https://github.com/kubernetes/kubernetes/pull/43987#issuecomment-297867514

File: Godeps/Godeps.json (hunk 2 above)
Comment (ericchiang) [reply Godeps/Godeps.json]: We import `gopkg.in/square/go-jose.v2` in `go-oidc` which I think is the recommended import. https://github.com/square/go-jose#versions

Maybe we can switch the import in k8s.io/kubernetes ?

Comment (mikedanese) [reply Godeps/Godeps.json]: Ya, my preference would be to update existing code to use gopkg.in... and avoid vendoring two of the same library.

Comment (easeway) [reply Godeps/Godeps.json]: Gotcha! I can check the places using go-jose.

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/login.go (hunk 1 above)
Comment (easeway) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/login.go]: Yes. I agree the implementation will be mostly the same. My implementation goes in the OIDC auth plugin in kubectl, so that will be OIDC specific. Regarding OAuth2, it's a different route performed by kubectl. And currently there's no specific design for an interactive flow. I feel we need more discussion on that before we can consolidate the implementation for both flows.

**Comment by easeway** [comment]
@pwittrock I will submit a design proposal soon. The issue already exists #54201

**Comment by ericchiang** [comment]
There was some discussion of this in sig-auth today about this. https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/view#heading=h.8jbewrjpj9jq

Biggest concerns are:

* This login mechanism should work with any rotation mechanism.
* Auth providers need to be pluggable and we should have a plan to add an extension mechanism before adding more.
  * Use cases include kubelet bootstrapping and the many other kind of IdPs we should support.
* The login mechanism should probably work with generic OAuth2.
* Syncing with sig-cli to determine what kind of things are okay to compile into kubectl.

Basically, before we start committing to an OpenID Connect login, we probably want to do some planning around client auth providers in general.

I'd like to take on some of this planning over the 1.10 release cycle.

**Comment by ericchiang** [comment]
Can people interested in this PR take a look at my proposal for external kubectl auth providers? Specifically the "Login" section?

https://github.com/kubernetes/community/pull/1503

**Comment by dims** [comment]
@easeway let's please close this. please reopen if necessary

/close
